### PR TITLE
[SPARK-30053][SQL] Add the ability for v2 datasource so specify a vacuum action on the table

### PR DESCRIPTION
**Pull request by AndrewKL**
https://issues.apache.org/jira/browse/SPARK-30053?page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel&focusedCommentId=16985760

### What changes were proposed in this pull request?

This cr includes a new api so that users of the V2 Datasource api can specify an action to take when a user types in the 'Vacuum {table}' command.  This is commonly used to remove deleted records and optimize the table.

### Why are the changes needed?

There is currently no way for users to create this type of behavior in SQL

> vacuum is a very common command. for example .
>
> - https://www.sqlite.org/lang_vacuum.html
> - https://www.postgresql.org/docs/9.1/sql-vacuum.html
> - https://docs.aws.amazon.com/redshift/latest/dg/r_VACUUM_command.html
> - https://docs.databricks.com/spark/latest/spark-sql/language-manual/vacuum.html
> 
> Another option would the optimize key word which databricks uses for delta.
> - https://docs.databricks.com/spark/latest/spark-sql/language-manual/optimize.html

### Does this PR introduce any user-facing change?

Yes this includes a new public but experimental API

### How was this patch tested?
 Unit tests.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
Just 2 cents as I'm not an expert of SQL area as well as just a contributor:

1) I read through the JIRA issue and see VACUUM is being supported for some systems. But do you have any custom data source which requires this, and if you have one could you please elaborate the plan? Without actual use case I'm not sure it's being accepted.

2) Looks like there're some Scala style issues. Could you please read through this doc https://github.com/databricks/scala-style-guide which is applied to Apache Spark repo as well? There're not only the lint error but also indentation issue which linter doesn't catch.

3) There're some spots which are modified for testing but they seem weird and shouldn't do it only for testing. I'll also leave comments.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 1]
```diff
@@ -1829,3 +1858,21 @@ class FakeV2Provider extends TableProvider {
...
+
+    override def name(): String = "vacuum"
+
+    override def schema(): StructType =
+      StructType(Seq(StructField("id", IntegerType)))
+
+    override def capabilities(): util.Set[TableCapability] =
+      Set(TableCapability.ACCEPT_ANY_SCHEMA).asJava
+
+    override def vacuum(): Unit = {println("VACUUM!!")}
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala]: 1. Where is the usage of this class?
2. Don't use println unless there's clear reason to do so. Use logXXX instead.
3. You may want to add flag here instead of modifying InMemoryTable. Please revert the change of InMemoryTable as it doesn't need to be modified.

**Review by HeartSaVioR — COMMENTED** [review]
Just general comments.

**Comment by AndrewKL** [comment]
> * I read through the JIRA issue and see VACUUM is being supported for some systems. But do you have any custom data source which requires this, and if you have one could you please elaborate the plan? Without actual use case I'm not sure it's being accepted.
> *

We have a custom Datasource where users can "DELETE" records from the table.  Internal these records are tomb stoned, instead of actually deleted.  This is a common design pattern in many relation table storage formats.

https://en.wikipedia.org/wiki/Tombstone_(data_store)

For GDPR compliance users would like to be able to force the cleanup process instead of waiting on an automated system to clean things up.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala (hunk 1 above)
Comment (AndrewKL) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala]: 1. This is a test class.
2/3. replaced this with a flag that is set.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/InMemoryTable.scala [hunk 2]
```diff
@@ -163,6 +164,12 @@ class InMemoryTable(
   override def deleteWhere(filters: Array[Filter]): Unit = dataMap.synchronized {
     dataMap --= InMemoryTable.filtersToKeys(dataMap.keys, partFieldNames, filters)
   }
+
+  var vacuumed = false
```
[the remark this answers is not in the document]
Comment (AndrewKL) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/connector/InMemoryTable.scala]: I'll create another purpose built connector.

**Comment by HyukjinKwon** [comment]
I copied and pasted the references mentioned in the JIRA into this PR description.

**Comment by HeartSaVioR** [comment]
> We have a custom Datasource where users can "DELETE" records from the table. Internal these records are tomb stoned, instead of actually deleted. This is a common design pattern in many relation table storage formats.

Yeah that's valid one. Many storages provide similar features, like `optimize table`, `compaction`, etc. I just wanted to confirm that you're proposing a new feature with real use cases.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsVacuum.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+ * interface to provide the ability to perform table maintenance on request of the user.
+ */
+@Experimental
+public interface SupportsVacuum {
+
+  /**
+   * Performs maintenance on the table.  This often includes removing unneeded data and
+   * deleting stale records.
+   *
+   * @throws IllegalArgumentException If the vacuum is rejected due to required effort.
```
Comment (HeartSaVioR) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsVacuum.java]: Previous comment seems to be go away so adding again (please avoid force pushing unless there's some conflict):

Throwing IllegalArgumentException sounds really weird if there's no argument. IMHO that should be some exception (even a new class) clearly representing the intention.

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
Thank you for contribution, @AndrewKL .
Please run `./dev/lint-scala` and fix the scala style?

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 4]
```diff
@@ -391,6 +391,7 @@ dmlStatementNoWith
     : insertInto queryTerm queryOrganization                                       #singleInsertQuery
     | fromClause multiInsertQueryBody+                                             #multiInsertQuery
     | DELETE FROM multipartIdentifier tableAlias whereClause?                      #deleteFromTable
+    | VACUUM multipartIdentifier tableAlias                                        #vacuumFromTable
```
Comment (rdblue) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Why add the table alias?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsVacuum.java (hunk 3 above)
Comment (rdblue) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsVacuum.java]: Why allow a source to reject a vacuum operation?

If vacuuming is supported, then the source should always do the operation when requested. I don't see a reason why the source would need to reject it.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsVacuum.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package org.apache.spark.sql.connector.catalog;
+
+import org.apache.spark.annotation.Experimental;
+/**
```
Comment (rdblue) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsVacuum.java]: Nit: missing blank line.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala [hunk 6]
```diff
@@ -304,8 +304,15 @@ case class DescribeTable(table: NamedRelation, isExtended: Boolean) extends Comm
  * The logical plan of the DELETE FROM command that works for v2 tables.
  */
 case class DeleteFromTable(
-    table: LogicalPlan,
-    condition: Option[Expression]) extends Command with SupportsSubquery {
+                            table: LogicalPlan,
```
[the remark this answers is not in the document]
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala]: I agree. These should be indented 4 spaces from where `case` started.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsVacuum.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+ * A mix-in interface for {@link Table} vacuum support. Data sources can implement this
+ * interface to provide the ability to perform table maintenance on request of the user.
+ */
+@Experimental
+public interface SupportsVacuum {
+
+  /**
+   * Performs maintenance on the table.  This often includes removing unneeded data and
```
Comment (rdblue) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsVacuum.java]: Should this be "unneeded metadata" instead?

I think that the behavior should be that vacuum does not modify the table's contents. DML operations should be used for that. Vacuum only cleans up metadata for faster operations, like an OPTIMIZE.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/InMemoryTable.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (rdblue) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/connector/InMemoryTable.scala]: I'm okay with this for tests. You could also keep track of something (like a list of former operations) and clean that up using VACUUM.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala [hunk 8]
```diff
@@ -177,6 +177,15 @@ object DataSourceV2Strategy extends Strategy with PredicateHelper {
           throw new AnalysisException("DELETE is only supported with v2 tables.")
       }
 
+    case VacuumTable(relation) =>
+      relation match {
+        case DataSourceV2ScanRelation(table, _, output) =>
+          VacuumTableExec(table.asVacuumable) :: Nil
+        case _ =>
+          throw new AnalysisException("VACUUM is only supported with v2 tables.")
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala]: This isn't the right place to throw an `AnalysisException` because this is after analysis has finished.

This rule should match `VacuumTable(DataSourceV2ScanRelation(table, _, _))` and produce the exec node. Then `CheckAnalysis` should have a rule that matches `VacuumTable` with any other relation that throws the analysis exception.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala (hunk 1 above)
Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala]: If this was replaced, can you remove it?

**Comment by rdblue** [comment]
@AndrewKL, this looks pretty close to me. Thanks for working on it!

@cloud-fan FYI.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 4 above)
Comment (AndrewKL) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Removed.

**Comment by cloud-fan** [comment]
shall we add an OPTIONS clause to be more flexible? in case people want some fine-grained control like still keeping data of the last 24 hours so that they can roll back if needed.

**Comment by HeartSaVioR** [comment]
@AndrewKL Do you have any update on this?
